Don't reject views with result types which are TypeVars

milessabin opened the pull request:
On the `matchesPtInst` fast path views are pruned without first being applied. This can result in a false negative in `HasMethodMatching` if the view has a result type which is a not fully instantiated `TypeVar`. The fix is to fall back to the slow path in that case.

Fixes scala/bug#11174.

milessabin committed:
Don't reject views with result types which are TypeVars

On the matchesPtInst fast path views are pruned without first being
applied. This can result in a false negative in HasMethodMatching if the
view has a result type which is a not fully instantiated TypeVar. The
fix is to fall back to the slow path in that case.

Fixes scala/bug#11174.

adriaanm commented on src/compiler/scala/tools/nsc/typechecker/Implicits.scala:
```diff
@@ -566,23 +566,39 @@ trait Implicits {
+
+            if(isView || isViewLike) {
+              tpInstantiated match {
+                case MethodType(_, tv: TypeVar) if !tv.instValid =>
```
I think this check could be more general. What if it's a conversion like `def foo(x: From): x.To`? The result type would also be undetermined, right? Or it could be an intersection type of a type variable with some other type.

Could you add a few test cases along these lines and make sure we catch those as well?

adriaanm reviewed (dismissed):
Nice minimization, now let's generalize what could possibly go wrong a bit :-)

milessabin committed:
When matching views instantiate TypeVars to their bounds

Implicits to satisfy views are matched against a search template. To
match correctly against the template, TypeVars in the candidates type
are replaced by their upper bounds once those bounds have been solved as
far as possible against the template.

milessabin commented:
Trying to get to the bottom of this ...

The issue boils down to the way that views are searched for when we're trying to adapt to a member. What we search for is a function from the adaptee to a synthetic refinement of a wildcard `? { def name: ? }` constructed by `HasMember`. In this case we'll be looking for a view matching `Test.c.type => ? { def foo: ? }`.

When we call `matchesPt` attempting to match the type of `toComponentCtor` against this template we end up in `isSubType` via `normSubType`. Here everything goes fine until we compare the result type of the method (`?CT[Int]`, a not yet fully instantiated `TypeVar` with an upper bound of `Props`) against the result type of the template (`? { def foo: ? }`) . This fails because we end up trying to register `? { def foo: ? }` as a bound on `CT` and this is rejected because `CT` is higher kinded whereas `? { def foo: ? }` isn't. A variant of this test case which has a result type of kind `*` succeeds, I think incorrectly, because the attempt to register the result type of the template as a bound is accepted thanks to the kinds lining up (see `test/files/pos/t11174c.scala`).

The fix in my first commit skirts around this problem by bailing to the slow path straight away. The new fix continues with the fast path, but replaces any uninstantiated `TypeVars` in the candidate type by their upper bounds before calling `matchesPt`. Consequently when we reach the previously problematic result type comparison it's now `Props[Int]` against `? { def foo: ? }` which succeeds as expected.

I understand your concern about views of other forms, but I think that this really is an interaction between the mechanism for matching against view templates in the presence of type variables. I've added a test case along the lines you suggested anyway ... it passes, but it's not really relevant to the problem.

dwijnand commented on src/compiler/scala/tools/nsc/typechecker/Implicits.scala:
```diff
@@ -566,39 +566,45 @@ trait Implicits {
+      }
+
+      if (StatisticsStatics.areSomeColdStatsEnabled) statistics.incCounter(matchesPtInstCalls)
+        info.tpe match {
```
The indentation looks off here (almost looks like this block is dependant on `if StatisticsStatics.areSomeColdStatsEnabled`)

milessabin replied:
Yes there is. I'll deal with that in a separate commit.

adriaanm commented on src/compiler/scala/tools/nsc/typechecker/Implicits.scala:
```diff
@@ -571,14 +571,36 @@ trait Implicits {
         case _ => false
       }
 
+      object tvarToHiBoundMap extends TypeMap {
```
Approximating a type variable by its upper bound is only sound when it occurs in a covariant position, right? See `ExistentialExtrapolation` for an analogous type map.

adriaanm replied:
(Also, `wildcardExtrapolation`)

milessabin replied:
That occurred to me too, but see also `dominates`: https://github.com/scala/scala/blob/2.13.x/src/compiler/scala/tools/nsc/typechecker/Implicits.scala#L455-L456.

adriaanm replied:
Yeah, that's not strictly right either, but has less impact since it's not about conformance.

adriaanm reviewed (dismissed):
I like the direction, but I'm not sure we're there yet. How much is there to gain in pruning polymorphic views? Perhaps we can just keep it simple and take the slow path for them?

adriaanm commented:
So, this is intriguing :-) I ran this simplified example through the debugger, and it looks like an issue with type constructor variables

```
trait SomeTC[TC[_]]

class Tycon[T] { def foo(props: T): T = ??? }

class DontPruneTyconView {
  implicit def toTycon[TC[_]](c: SomeTC[TC]): TC[Int] = ???

  (??? : SomeTC[Tycon]).foo(23)
}

```

<img width="1126" alt="screen shot 2018-10-03 at 2 57 18 pm" src="https://user-images.githubusercontent.com/91083/46411681-aff6d100-c71c-11e8-84e1-1b4abf5efaad.png">

adriaanm commented:
Concretely, there's no case in registerBound to deal with the situation where an applied typeconstructor variable is compared against a bound that's not also a type constructor. It makes perfect sense to register the bound `[T >: Int <: Int] ?{def foo: ?}` on the `?TC` type var, for its occurrence in the type application `?TC[Int]`.

milessabin commented:
Yup, that's what I was trying to say earlier ;-)

> This fails because we end up trying to register ? { def foo: ? } as a bound on CT and this is rejected because CT is higher kinded whereas ? { def foo: ? } isn't. 

But I think we shouldn't be trying to do this in the first place, hence substituting with the bound.

How about if I make the replacement respect variance?

adriaanm commented:
Apologies, I missed that in your comment.

adriaanm commented:
I do think that missing case is a bug. We should allow a type constructor var that's applied to type arguments (thus of kind *) to register kind-* bounds. In the implementation, this tracking has to be done "inside out" -- sketching a solution:

```
      def unifyApplied(tp: Type) = this match {
        case atv: AppliedTypeVar =>
          addBound(genPolyType(params, tp))
          true
        case _ => false
      }
```

and patch this in here: `        unifySimple || unifyFull(tp) || unifyApplied(tp) || (`

milessabin commented:
If you try with `test/files/pos/t11174c.scala`, which is an example with a result type of kind `*`, you'll see that the bound registration _does_ go through. But `? { def foo: ? }` doesn't really make sense as a bound ... as I said, I don't think we should be trying this with uninstantiated `TypeVars`.

adriaanm commented:
Why does it not make sense as a bound?

adriaanm commented:
Also, the other case is routinely comparing type vars against other types, so I'm not sure I follow the other point.

milessabin commented:
> Why does it not make sense as a bound?

Well, maybe it does.

But I'm slightly puzzled that `{ def foo(props: T): T }` can be a subtype of `? { foo: ? }` given that `foo` has no argument list in the latter. Maybe that's not a problem given the role that it's playing here.

adriaanm commented:
The syntax is confusing, but it just means there's a member named `foo` of any type, which subsumes a val or a method of any arity (it would have a MethodType as info that encodes its argument list, but we leave the info as a wildcard)

adriaanm commented:
The refinement type is created by `HasMember`

adriaanm commented:
Btw, there's a very similar test already in our suite as test/files/pos/t7785.scala

adriaanm commented:
All this aside, it's most likely too risky to change type inference to fix this. My curiosity just got piqued...

milessabin commented:
> All this aside, it's most likely too risky to change type inference to fix this.

Agreed for 2.12.x. But what about for 2.13.x?

milessabin commented:
How do you want to proceed from here? Fix the variance issue with the "replace by bounds" version?

adriaanm commented:
I don't know -- it's tricky :-) I'm worried about regressions, but also concerned with the asymmetry between views and regular implicit search. The type variable bound bug we both discovered could also bite in a regular implicit search.

milessabin commented:
This non-view looks comparable and isn't problematic,

```scala
object Test {
  implicit def mkF[F[_]]: F[Int] = ???
  implicitly[Option[Int]]
}
```

adriaanm commented:
This uses the existing unification mechanisms. I'm looking for an example that decides that `?F[Int]` could not plausibly be a subtype (doesn't unify) with some kind-* type which in fact could be a supertype.

adriaanm commented:
Given the complexity of it all, I would be inclined to just bail on this for 2.12.8 and use the slow path for polymorphic views.

milessabin commented:
> Given the complexity of it all, I would be inclined to just bail on this for 2.12.8 and use the slow path for polymorphic views.

OK, I'll do that (and fix the whitespace bug in a separate commit).

milessabin committed:
Take views off the fast path

milessabin commented:
I've taken views off the fast path, and also tentatively removed the `isViewLike` logic as well. I needed that to get the community build passing previously, but I suspect that bailing on all views might make it unnecessary. If a community build run fails on it then it can be reinstated.

joroKr21 commented:
> This uses the existing unification mechanisms. I'm looking for an example that decides that ?F[Int] could not plausibly be a subtype (doesn't unify) with some kind-* type which in fact could be a supertype.

May be related to #6573

milessabin commented:
Community build run in progress [here](https://scala-ci.typesafe.com/view/scala-2.12.x/job/scala-2.12.x-integrate-community-build/3610/).

milessabin commented:
The community build is _almost_ green. The two failures are classutil and enumeratum.

The classutil failure is due to a runtime test that the Scala release version is 2.12.6. So, unrelated to this PR or implicit pruning.

I'm not quite sure what the enumeratum failure is, but it's runtime again and doesn't seem obviously related to this PR or implicit pruning in general. The enumeratum build seems quite flaky and I'm not having a lot of luck getting it to build locally even with vanilla 2.12.7 due to Scala.js version slippage. 

@SethTisue any thoughts on that?

SethTisue commented:
agree w/ your & Adriaan's assessment of the results: it's fine

milessabin commented:
@adriaanm is this good to go?

It'll need carrying over to 2.13.x ... I'd be happy to look at making the fix in `TypeVar` if that works for you.

adriaanm commented:
It would be good to investigate a better fix, but I'm not sure it'll be safe enough to include in 2.13. I guess it depends on the solution, but 2.13 at this point isn't much more permissive of regressions than 2.12. I don't want to distract from the focus of polish and stability for the collections.
